cli/sub_command.go

```diff
@@ -139,89 +145,56 @@ func (c *subCmd) subscribe(_ *fisk.ParseContext) error {
 		}
 
 		ctr++
-		if dump {
-			// Output format 1/3: dumping, to stdout or files
-
-			stdout := c.dump == "-"
-			outFile := ""
-			if !stdout {
-				if info == nil {
-					outFile = filepath.Join(c.dump, fmt.Sprintf("%d.json", ctr))
-				} else {
-					outFile = filepath.Join(c.dump, fmt.Sprintf("%d.json", info.StreamSequence()))
-				}
-			}
-
-			// dont want sub etc
-			msg := nats.NewMsg(m.Subject)
-			msg.Header = m.Header
-			msg.Data = m.Data
-			msg.Reply = m.Reply
-
-			jm, err := json.Marshal(msg)
-			if err != nil {
-				log.Printf("Could not JSON encode message: %s", err)
-			} else if stdout {
-				os.Stdout.WriteString(fmt.Sprintf("%s\000", jm))
-			} else {
-				err = os.WriteFile(outFile, jm, 0600)
-				if err != nil {
-					log.Printf("Could not save message: %s", err)
-				}
-
-				if ctr%100 == 0 {
-					fmt.Print(".")
-				}
-			}
-
-		} else if c.raw {
-			// Output format 2/3: raw
-
-			fmt.Println(string(m.Data))
-
+		if c.match && m.Reply != "" {
+			matchMap[m.Reply] = m
 		} else {
-			// Output format 3/3: pretty
+			printMsg(c, m, nil, ctr)
+		}
 
-			if info == nil {
-				if m.Reply != "" {
-					fmt.Printf("[#%d] Received on %q with reply %q\n", ctr, m.Subject, m.Reply)
-				} else {
-					fmt.Printf("[#%d] Received on %q\n", ctr, m.Subject)
-				}
-			} else if jetStream {
-				fmt.Printf("[#%d] Received JetStream message: stream: %s seq %d / subject: %s / time: %v\n", ctr, info.Stream(), info.StreamSequence(), m.Subject, info.TimeStamp().Format(time.RFC3339))
-			} else {
-				fmt.Printf("[#%d] Received JetStream message: consumer: %s > %s / subject: %s / delivered: %d / consumer seq: %d / stream seq: %d\n", ctr, info.Stream(), info.Consumer(), m.Subject, info.Delivered(), info.ConsumerSequence(), info.StreamSequence())
+		if ctr == c.limit {
+			sub.Unsubscribe()
+			// if no reply matching, or if didn't yet get all replies
+			if !c.match || len(matchMap) == 0 {
+				cancel()
 			}
+		}
+	}
 
-			if len(m.Header) > 0 {
-				for h, vals := range m.Header {
-					for _, val := range vals {
-						fmt.Printf("%s: %s\n", h, val)
-					}
-				}
-
-				fmt.Println()
-			}
+	matchHandler := func(reply *nats.Msg) {
+		mu.Lock()
+		defer mu.Unlock()
 
-			if !c.headersOnly {
-				fmt.Println(string(m.Data))
-				if !strings.HasSuffix(string(m.Data), "\n") {
-					fmt.Println()
-				}
-			}
+		request, ok := matchMap[reply.Subject]
+		if !ok {
+			return
+		}
 
-		} // output format type dispatch
+		printMsg(c, request, reply, ctr)
+		delete(matchMap, reply.Subject)
 
-		if ctr == c.limit {
-			sub.Unsubscribe()
+		// if reached limit and matched all requests
```

Comment:
If we do decide to not count these 2 the message count we can just remove this here

Reply:
When we got `count` requests, we still have to wait for a response to any unmatched request